### PR TITLE
Update doc because local doesn't mean local-returning

### PR DESCRIPTION
**Pull request by riaqn**
This finishes the work in #2669

### CODE REVIEW

File: ocaml/jane/doc/extensions/local/reference.md [hunk 1]
```diff
@@ -460,11 +460,8 @@ value `x` remains available.
 
 The region around the body of a function prevents local allocations inside that
 function from escaping. Occasionally, it is useful to write a function that
-allows local allocations to escape, which can be done by explicitly marking
-such functions.
-
-This is useful particularly for constructor functions of abstract types. For
-instance, consider this code that uses an `int ref` as a counter:
+allocates and returns value in caller's region. For instance, consider this code
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:463]:
```suggestion
allocates and returns a value in caller's region. For instance, consider this code
```

File: ocaml/jane/doc/extensions/local/reference.md [hunk 2]
```diff
@@ -502,46 +499,38 @@ let f () =
...
-allocated. This remains the case no matter how many local_ annotations we write
-inside `f`: the issue is the definition of `make`, not its uses.
+allocated. This remains the case no matter how many `local_` annotations we
+write inside `f`: the issue is the definition of `make`, not its uses.
 
-To allow the counter to be locally allocated, we need to specify that
-`Counter.make` may return local allocations. This can be done by wrapping the
-entire body of `make` with the `local_` keyword:
+To allow the counter to be locally allocated, we need to make `Counter.make`
+allocate and return in caller's region. This can be done by `exclave_`:
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:506]:
```suggestion
end its region early so that it can allocate its return value in the caller's region. This can be done with `exclave_`:
```

File: ocaml/jane/doc/extensions/local/reference.md [hunk 3]
```diff
@@ -502,46 +499,38 @@ let f () =
...
-
-Instead, local allocations during `make` are considered part of `f`s region,
-and will only be cleaned up when that region ends. Local allocations are
-allocated as always in the nearest enclosing region. However if the current
-function is a local-returning function, then the nearest enclosing region will
-be the caller's (or that of the caller's caller, etc., if the caller is also
-local-returning).
+The keyword `exclave_` terminates the current region and executes the subsequent
+code in the outer region. Therefore, `ref 0` is executed in `f`'s region, which
+allows its local allocation. The allocation will only be cleaned up when the
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:515]: Note to self: when we merge this PR, we should make a follow-on PR that uses the term "stack allocation" consistently instead of "local allocation".

File: ocaml/jane/doc/extensions/local/reference.md [hunk 4]
```diff
@@ -502,46 +499,38 @@ let f () =
...
+region of `f` ends.
 
-## Exclave
-In the previous section, we discussed that a function can return local values
-without having its own region. Consequently, it operates within the caller's
-region. This approach, however, has certain disadvantages. Consider the
-following example:
+## Advanced usage of exclaves
+In the previous section, the example function exits its own region immediately,
+which allows allocating and returning in caller's region. This approach,
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:520]:
```suggestion
which allows allocating and returning in the caller's region. This approach,
```

File: ocaml/jane/doc/extensions/local/reference.md [hunk 5]
```diff
@@ -550,18 +539,17 @@ let f (local_ x) =
...
-terminated by `exclave_`, releasing all temporary allocations. Both `None` and
-`Some x` are considered "local" relative to the outer region and are allowed to
-escape. In summary, we have temporary allocations on the stack that are promptly
-released and result allocations on the stack that can escape.
+In this example, the function `f` has a region where the allocation for the
+complex computation occurs. This region is terminated by `exclave_`, releasing
+all temporary allocations. Both `None` and `Some x` are considered "local"
+relative to the outer region and are allowed to be returned. In summary, we have
+temporary allocations in the `f`'s region that are promptly released and result
+allocations in the caller's region that can be returned.
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:547]:
```suggestion
relative to the outer region and are allowed to be returned. In summary, the
temporary allocations in `f`'s region are promptly released, and the result
allocations in the caller's region can be returned.
```

File: ocaml/jane/doc/extensions/local/reference.md [hunk 6]
```diff
@@ -579,11 +567,9 @@ This function is intended to have the type:
 val maybe_length : ('a -> bool) -> 'a list -> local_ int option
 ```
 It is designed not to allocate heap memory by using the stack for all `Some`
-allocations.  However, it will currently use O(N) stack space because all
+allocations. However, it will currently use O(N) stack space because all
 allocations occur in the original caller's stack frame. To improve its space
-usage, we remove the `local_` annotation (so the function has its own region),
-and wrap `Some (count + 1)` inside `exclave_` to release the region before the
-allocation:
+usage, we delay the `exclave_` annotation until returning result:
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:572]:
```suggestion
usage, we delay the `exclave_` annotation until returning the result:
```

File: ocaml/jane/doc/extensions/local/reference.md [hunk 7]
```diff
@@ -460,11 +460,8 @@ value `x` remains available.
```
Comment (ncik-roberts) [inline ocaml/jane/doc/extensions/local/reference.md:460]: The title of this section "local-returning functions" should be renamed to mention `exclave_`. I suggest:

## Use `exclave_` to return a stack allocation

Comment (ncik-roberts) [reply ocaml/jane/doc/extensions/local/reference.md:460]: This also requires updating references to this section, e.g. "See "Local-returning functions" below."

**Review by ncik-roberts — COMMENTED** [review]
Looks good. I'll take another quick glance before we merge.

File: ocaml/jane/doc/extensions/local/reference.md (hunk 3 above)
Comment (riaqn) [reply ocaml/jane/doc/extensions/local/reference.md:515]: Yes I think [#2540](https://github.com/ocaml-flambda/flambda-backend/pull/2540#discussion_r1653222006) is very related. My current plan is that that PR will be merged without document and notifying the user. We will (in a few weeks) do a complete overhaul of this document, talk about "stack allocation" and introduce the `stack_` syntax.

File: ocaml/jane/doc/extensions/local/reference.md (hunk 7 above)
Comment (riaqn) [reply ocaml/jane/doc/extensions/local/reference.md:460]: Yes I also thought about this, but thought maybe "local-returning functions" is what users want, and "exclave" is just a way (currently the only way) to achieve that. Let me know if you still think we shoud change to "exclave". (I have no preference).

Comment (ncik-roberts) [reply ocaml/jane/doc/extensions/local/reference.md:460]: I think that the either the title or the first line of the section should mention “exclave” to help users scan the doc.

maybe the first sentence approach is enough?

> TL;DR: Use `exclave_` to return a local value.

“Local value” is better than “local allocation.”

**Comment by riaqn** [comment]
Closing this PR, because I messed up the branch. Superseded by #2734

File: ocaml/jane/doc/extensions/local/reference.md (hunk 7 above)
Comment (riaqn) [reply ocaml/jane/doc/extensions/local/reference.md:460]: Thanks - I made the suggested change in #2734
